File: backend/models/Product.js
const mongoose = require('mongoose');

const reviewSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  name: { type: String, required: true },
  rating: { type: Number, required: true, min: 1, max: 5 },
  comment: { type: String },
}, { timestamps: true });

const productSchema = new mongoose.Schema({
  vendor: { type: mongoose.Schema.Types.ObjectId, ref: 'Vendor', required: true },
  name: { type: String, required: [true, 'Product name is required'], trim: true },
  slug: { type: String, unique: true },
  description: { type: String, required: [true, 'Description is required'] },
  price: { type: Number, required: [true, 'Price is required'], min: 0 },
  compareAtPrice: { type: Number, min: 0 },
  category: { type: String, required: [true, 'Category is required'], index: true },
  subcategory: { type: String },
  brand: { type: String },
  images: [{ type: String }],
  condition: {
    type: String,
    enum: ['new', 'like-new', 'open-box', 'refurbished', 'used'],
    default: 'open-box',
  },
  stock: { type: Number, required: true, min: 0, default: 0 },
  sku: { type: String },
  weight: { type: Number },
  dimensions: { length: Number, width: Number, height: Number },
  tags: [{ type: String }],
  reviews: [reviewSchema],
  averageRating: { type: Number, default: 0 },
  numReviews: { type: Number, default: 0 },
  isActive: { type: Boolean, default: true },
  isFeatured: { type: Boolean, default: false },
}, { timestamps: true });

productSchema.pre('save', function (next) {
  this.slug = this.name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
  if (this.reviews.length > 0) {
    this.averageRating = this.reviews.reduce((sum, r) => sum + r.rating, 0) / this.reviews.length;
    this.numReviews = this.reviews.length;
  }
  next();
});

productSchema.index({ name: 'text', description: 'text', tags: 'text' });

module.exports = mongoose.model('Product', productSchema);

File: ios/Packages/GOBShared/Sources/GOBShared/Models/Product.swift
import Foundation

public struct Product: Codable, Identifiable, Sendable, Hashable {
    public let id: String
    public let vendorId: String?
    public let vendorName: String?
    public let name: String
    public let slug: String?
    public let description: String
    public let price: Double
    public let compareAtPrice: Double?
    public let category: String
    public let subcategory: String?
    public let brand: String?
    public let condition: String
    public let stock: Int
    public let sku: String?
    public let images: [String]
    public let tags: [String]?
    public let rating: Double?
    public let numReviews: Int?
    public let isFeatured: Bool?

    public var discountPercent: Int? {
        guard let compareAt = compareAtPrice, compareAt > price else { return nil }
        return Int(((compareAt - price) / compareAt) * 100)
    }

    public var firstImageURL: URL? {
        guard let first = images.first else { return nil }
        return URL(string: first)
    }

    public static func == (lhs: Product, rhs: Product) -> Bool {
        lhs.id == rhs.id
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}
